### PR TITLE
Dictionaries

### LINKED ISSUE #63: Common Collections: Dictionaries
This chapter will be _extremely_ detailed, and will need provide an extensive description of how dictionaries work on the lower level, inspired by the current [Cairo docs](https://www.cairo-lang.org/docs/reference/common_library.html?highlight=dict_squash#default-dict-new) to describe why dicts need to be squashed.

Estimated time: 3 hours

### PR DESCRIPTION
**Pull request by rodrigo-pino**
Fixes #63 

Adds a new chapter explaining `DictFelt252<T>`.

Content wise:
- [x] Basic Dict functionality
- [x] Dictionaries Underneath
  - [x] Explain Entry List
  - [x] Explain squashing and why it is need
  - [x] Explain Destruct
- [x] Additional example use of Dicts
  - [x] Using the `entry` method
  - [x] How dictionaries interact with Structs
  - [ ] ~How dictionaries interact with Enums?~

Also it still lacks polishing, specially the later sections, and run it through a spell check.

### CODE REVIEW

**Comment by enitrat** [comment]
> How dictionaries interact with Structs
> How dictionaries interact with Enums?

I'm not sure about Enums, but for structs, you can create a Felt252Dict<Nullable<T>> to store your struct in the dict easily!

File: src/ch15-01-dictionaries.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+For each time we interact with a `Felt252Dict<T>` a new `Entry<T>` will be registered:
+
+- A `get` would register an entry where there is no change in state, and previous and new values are stored with the same value.
+- An `insert` would register new `Entry<T>` where the `new_value` would be the element being inserted, and the `previous_value` the last element inserted before this. In case it is the first entry for certain key, then the previous and current value will be the same.
+
+The use of this entry list shows how there isn't any rewriting, just the creation of new memory cells per `Felt252Dict<T>` interaction. Let's show an example of this using the `balances` dictionary from the previous section and inserting the users 'Alex' and 'Maria':
+
+```rust
+balances.insert('Alex', 100_u64);
```
Comment (enitrat) [inline src/ch15-01-dictionaries.md:96]: To reference this extract of cairo, you can create a `.cairo` file under listings and use the anchor system to import it here: https://rust-lang.github.io/mdBook/format/mdbook.html?highlight=ANCHOR#including-portions-of-a-file

File: src/ch15-01-dictionaries.md [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+|  key  | previous | current |
+| :---: | -------- | ------- |
+| Alex  | 100      | 100     |
+| Maria | 50       | 50      |
+| Alex  | 100      | 200     |
+| Maria | 50       | 50      |
+
+Notice that since 'Alex' was inserted twice, it appears twice and the `previous` and `current` values are set properly. Also reading from 'Maria' registered an entry with no change from previous to current values.
+
+This approach to implement `Felt252Dict<T>` means that for each reading/writing operations there is a scan for the whole entry list in search the last entry with the same `key`. Once the entry have been found, its `new_value` is extacted and used on the new entry to be added as the `previous_value`. This means that interacting with `Felt252Dict<T>` has a worst time complexity of `O(n)` where `n` is the amount of entries in the list.
```
Comment (enitrat) [inline src/ch15-01-dictionaries.md:113]: > This means that interacting with `Felt252Dict<T>` has a worst time complexity of `O(n)` where `n` is the amount of entries in the list.

Are we sure about this? How can we evaluate that?

File: src/ch15-01-dictionaries.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+|  Maria  | 150      | 190     |
+| Charles | 70       | 70      |
+
+In case of a change on any of the values of the first table, squashing would have failed during runtime.
+
+### Dictionary Destruction
+
+If you ran the examples from [Basic use of Dictionaries](##Basic use of Dictionaries) you'd probably notice that you never did a call to squash dictionary, but program compiled succesfully nonetheless. What happened behind the scene was that squash was called automatically via the `Felt252Dict<T>` implementation of the `Destruct<T>` trait. This call ocurred just before `balance` went out of scope.
+
+The `Destruct<T>` trait represents another way of removing instances out of scope apart from `Drop<T>`. The main difference between these two is that `Drop<T>` is treated as a no-op operation, meaning it does not generate new CASM while `Destruct<T>` does not have this restriction. The only type which actively uses the `Destruct<T>` trait is `Felt252Dict<T>`, for every other type `Destruct<T>` and `Drop<T>` are synonims.
```
Comment (enitrat) [inline src/ch15-01-dictionaries.md:152]: Can you add a reference to appendix-03 ### Drop and Destruct?

File: src/ch15-01-dictionaries.md [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+Up to this point we have given a comprehensive overview of the functionality of `Felt252Dict<T>` as well as how and why it is implemented in this way. If you haven't understood all of it, don't worry because in this section we will have some hands on examples using dictionaries in non-trivial ways.
+
+We will begin explaining the `entry` method which is part of a dictionary basic functionality included `Felt252DictTrait<T>` which we didn't mentioned at the beginning. After that we will see examples of how `Felt252Dict<T>` interacts with other types such as structs and enums.
+
+### The `entry` method
+
+Remember when we mentioned
+
+by using the entry function you can update the dictionary in a new way
```
Comment (enitrat) [inline src/ch15-01-dictionaries.md:166]: quotes

File: src/ch15-01-dictionaries.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+It returns a dict entry according to a key
+
```
Comment (enitrat) [inline src/ch15-01-dictionaries.md:169]: extra newline

File: src/ch15-01-dictionaries.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+which is a whole new type. Once you ask for an entry the original dict cannot be modified in any way until you finalize working with this entry
+
+Once you do such entry.finalize you are given back access to the normal dictionary
+
+There is no performance gain, but it represents another idiomatic way of modifying a dictionary.
+
```
Comment (enitrat) [inline src/ch15-01-dictionaries.md:175]: Can you show an example for this?

**Review by enitrat — COMMENTED** [review]
Incredible work @rodrigo-pino, it's super clear and well written 🔥

**Comment by rodrigo-pino** [comment]
I added the core of what I wanted to say in the last section. On monday I'll give it another pass to improve coherence and typo correction. Also to use the listings for the code.

I wanted to reference the Nullable in the book, but I don't think it is written yet, so I added a small paragraph explaining it but it may be incomplete, or not entirely correct since I haven't researched it in depth as for now.

File: src/ch15-01-dictionaries.md (hunk 2 above)
Comment (rodrigo-pino) [reply src/ch15-01-dictionaries.md]: Uff I gather that after diving into the sierra to CASM compiler, and seen the generated code. It was a loop through the memory segment comparing keys until the last one. It was some time a go, I'll recheck and  try to link the file where I saw it

File: src/ch15-01-dictionaries.md (hunk 4 above)
Comment (rodrigo-pino) [reply src/ch15-01-dictionaries.md]: This was the part that I hadn't wrote 🙈 , just random ideas to link later. The remaining comments address unwritten stuff as well

File: src/ch15-01-dictionaries.md (hunk 2 above)
Comment (enitrat) [reply src/ch15-01-dictionaries.md]: Awesome, thanks! 🙌

File: src/ch15-01-dictionaries.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+Until this point we have seen all the basic features of `Felt252Dict<T>` and how its behaviour mimics the same behaviour as the corresponding data structures in any other language, that is, externally of course. Cairo is at its core a non-deterministic Turing-complete programming language, very different from any other popular language in existence, which as a consequence means that dictionaries are implemented very different as well!
+
+In the following sections we are going to give some insights about a Dictionary inner mechanisms and the compromises that were taken to make them work. After that we are going to take a look at how to use dictionaries with other data structures as well as other ways of interacting with it.
+
+## Dictionaries Underneath
+
+One of the constraints of being a non-deterministic language is that the memory system is inmutable, so in order to simulate mutability, the team decided to implement `Felt252Dict<T>` as list of entries. Each of the entries represent a time where a dictionary was accessed for reading/updating/writing purposes. An entry have three fields:
```
[the remark this answers is not in the document]
Comment (rodrigo-pino) [reply src/ch15-01-dictionaries.md]: 👍

**Review by enitrat — APPROVED** [review]
Awesome!!!

**Pull request merged**